### PR TITLE
Unify error messages and add Troubleshooting section to docs

### PR DESCRIPTION
**Pull request by tjzel**
## Summary

Since version checker PR #4914 will add more error messages to Reanimated and our Troubleshooting section in docs was almost non-existent I decided to add a new one.

To make it consistent I made all errors thrown by Reanimated start with `[Reanimated]` and prettified them a bit.

With this I made this tiny script `validate-urls.js` that scans our source code to get valid (not so exotic) URLs and checks if they are available.

## Test plan

CI

### CODE REVIEW

File: android/src/main/java/com/swmansion/reanimated/layoutReanimation/AnimationsManager.java [hunk 1]
```diff
@@ -410,7 +410,8 @@ private static void addProp(WritableMap propMap, String key, Object value) {
     } else if (value instanceof ReadableMap) {
       propMap.putMap(key, (ReadableMap) value);
     } else {
-      throw new IllegalStateException("Unknown type of animated value [Layout Animations]");
+      throw new IllegalStateException(
+          "[Reanimated] Unknown type of animated value [Layout Animations].");
```
Comment (tomekzaw) [inline android/src/main/java/com/swmansion/reanimated/layoutReanimation/AnimationsManager.java:414]:
```suggestion
          "[Reanimated] Unknown type of animated value.");
```

File: docs/docs/guides/troubleshooting.mdx [hunk 2]
```diff
@@ -0,0 +1,110 @@
+---
+id: Troubleshooting
+title: Troubleshooting
+sidebar_label: Troubleshooting
+sidebar_position: 1
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:5]:
```suggestion
sidebar_position: 999
```

File: docs/docs/guides/troubleshooting.mdx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+---
+
+# Troubleshooting
+
+## Version mismatching
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:10]:
```suggestion
## Initialization issues
```

File: docs/docs/guides/troubleshooting.mdx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+- C++,
+- Java,
+- JavaScript,
+- Reanimated Babel plugin.
+
+All of them are supposed to work correctly only within the same minor version. Therefore, having any of those pieces in your code - directly or indirectly - separate from `react-native-reanimated`, especially having any code transpiled with a different version of the aforementioned plugin will result in undefined behavior and errors.
+
+### Unknown version of Reanimated Babel plugin
+
+This can happen when:
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:23]:
```
**Problem:** Lorem ipsum sit dolor amet.

**Solution:** Lorem ipsum sit dolor amet.
```

File: src/reanimated2/Bezier.ts [hunk 5]
```diff
@@ -97,7 +97,7 @@ export function Bezier(
   }
 
   if (!(mX1 >= 0 && mX1 <= 1 && mX2 >= 0 && mX2 <= 1)) {
-    throw new Error('bezier x values must be in [0, 1] range');
+    throw new Error('[Reanimated] bezier x values must be in [0, 1] range');
```
Comment (tomekzaw) [inline src/reanimated2/Bezier.ts:100]:
```suggestion
    throw new Error('[Reanimated] Bezier x values must be in [0, 1] range.');
```

File: src/reanimated2/NativeReanimated/NativeReanimated.ts [hunk 6]
```diff
@@ -71,12 +71,11 @@ export class NativeReanimated {
       const { ReanimatedModule } = NativeModules;
       ReanimatedModule?.installTurboModule();
     }
+
```
Comment (tomekzaw) [inline src/reanimated2/NativeReanimated/NativeReanimated.ts:74]:
```suggestion
```

File: src/reanimated2/animation/transformationMatrix/matrixUtils.tsx [hunk 7]
```diff
@@ -345,7 +345,7 @@ export function decomposeMatrix(
 
   // normalize matrix
   if (matrix[15] === 0) {
-    throw new Error('Invalid transform matrix!');
+    throw new Error('[Reanimated] Invalid transform matrix!');
```
Comment (tomekzaw) [inline src/reanimated2/animation/transformationMatrix/matrixUtils.tsx:348]:
```suggestion
    throw new Error('[Reanimated] Invalid transform matrix.');
```

File: src/reanimated2/animation/util.ts [hunk 8]
```diff
@@ -55,7 +55,7 @@ function recognizePrefixSuffix(value: string | number): RecognizedPrefixSuffix {
     );
     if (!match) {
       throw Error(
-        "Couldn't parse animation value. Check if there isn't any typo."
+        "[Reanimated] Couldn't parse animation value. Check if there isn't any typo."
```
Comment (tomekzaw) [inline src/reanimated2/animation/util.ts:58]:
```suggestion
        "[Reanimated] Couldn't parse animation value."
```

File: src/reanimated2/core.ts [hunk 9]
```diff
@@ -50,7 +50,7 @@ global._log = function (s: string) {
 export function getViewProp<T>(viewTag: number, propName: string): Promise<T> {
   if (global._IS_FABRIC) {
     throw new Error(
-      '[react-native-reanimated] `getViewProp` is not supported on Fabric yet'
+      '[Reanimated] `getViewProp` is not supported on Fabric yet'
```
Comment (tomekzaw) [inline src/reanimated2/core.ts:53]:
```suggestion
      '[Reanimated] `getViewProp` is not supported on Fabric yet.'
```

File: src/reanimated2/globals.d.ts [hunk 10]
```diff
@@ -22,6 +22,8 @@ declare global {
   var _REANIMATED_IS_REDUCED_MOTION: boolean | undefined;
   var _IS_FABRIC: boolean | undefined;
   var _REANIMATED_VERSION_CPP: string | undefined;
+  var _REANIMATED_VERSION_JAVA: string | undefined;
```
Comment (tomekzaw) [inline src/reanimated2/globals.d.ts:25]: w innym

File: src/reanimated2/interpolation.ts [hunk 11]
```diff
@@ -78,7 +78,7 @@ function validateType(type: ExtrapolationType): RequiredExtrapolationConfig {
   if (typeof type === 'string') {
     if (!isExtrapolate(type)) {
       throw new Error(
-        `Reanimated: not supported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
+        `[Reanimated] Not supported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```
Comment (tomekzaw) [inline src/reanimated2/interpolation.ts:81]:
```suggestion
        `[Reanimated] Unsupported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```

File: src/reanimated2/interpolation.ts [hunk 12]
```diff
@@ -93,7 +93,7 @@ function validateType(type: ExtrapolationType): RequiredExtrapolationConfig {
     (type.extrapolateRight && !isExtrapolate(type.extrapolateRight))
   ) {
     throw new Error(
-      `Reanimated: not supported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
+      `[Reanimated] Not supported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```
Comment (tomekzaw) [inline src/reanimated2/interpolation.ts:96]:
```suggestion
      `[Reanimated] Unsupported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```

File: src/reanimated2/layoutReanimation/animationBuilder/BaseAnimationBuilder.ts [hunk 13]
```diff
@@ -16,7 +16,7 @@ export class BaseAnimationBuilder {
   ) => InstanceType<T>;
 
   build = (): EntryExitAnimationFunction | LayoutAnimationFunction => {
-    throw Error('Unimplemented method in child class.');
+    throw Error('[Reanimated] Unimplemented method in child class.');
```
Comment (tomekzaw) [inline src/reanimated2/layoutReanimation/animationBuilder/BaseAnimationBuilder.ts:19]:
```suggestion
    throw new Error('[Reanimated] Unimplemented method in child class.');
```

File: src/reanimated2/layoutReanimation/animationBuilder/Keyframe.ts [hunk 14]
```diff
@@ -65,7 +65,7 @@ class InnerKeyframe implements IEntryExitAnimationBuilder {
     */
     if (!this.definitions['0']) {
       throw Error(
-        "Please provide 0, or 'from' keyframe with initial state of your object."
+        "[Reanimated] Please provide 0, or 'from' keyframe with initial state of your object."
```
Comment (tomekzaw) [inline src/reanimated2/layoutReanimation/animationBuilder/Keyframe.ts:68]:
```suggestion
        "[Reanimated] Please provide 0 or 'from' keyframe with initial state of your object."
```

File: android/build.gradle [hunk 15]
```diff
@@ -258,7 +256,7 @@ android {
         targetSdkVersion safeExtGet("targetSdkVersion", 30)
         versionCode 1
         versionName "1.0"
-        buildConfigField "boolean", "IS_NEW_ARCHITECTURE_ENABLED", IS_NEW_ARCHITECTURE_ENABLED.toString()
+        buildConfigField("boolean", "IS_NEW_ARCHITECTURE_ENABLED", IS_NEW_ARCHITECTURE_ENABLED.toString())
```
Comment (tomekzaw) [inline android/build.gradle:259]: I know this syntax is not really intuitive but we use it quite consistently here so probably there's no need to make this change.

File: android/build.gradle [hunk 16]
```diff
@@ -449,9 +442,7 @@ def assertLatestReactNativeWithNewArchitecture = task assertLatestReactNativeWit
     onlyIf { IS_NEW_ARCHITECTURE_ENABLED && REANIMATED_MAJOR_VERSION == 3 && REACT_NATIVE_MINOR_VERSION < 72 }
     doFirst {
         throw new GradleException(
-            "\n[react-native-reanimated] Reanimated " + REANIMATED_VERSION + " supports the New Architecture " +
-            "only on the latest minor release of React Native. Please upgrade to React Native 0.72.0+ " +
-            "or downgrade to an older version of Reanimated v3."
+            "\n[Reanimated] Outdated version of `react-native` for New Architecture. Reanimated " + REANIMATED_VERSION + " supports the New Architecture on React Native 0.72.0+."
```
Comment (tomekzaw) [inline android/build.gradle:445]: Why did you change this error message? Can we add a link to the Troubleshooting section here?

File: android/src/main/cpp/NativeProxy.cpp [hunk 17]
```diff
@@ -508,7 +507,8 @@ void NativeProxy::setupLayoutAnimations() {
                 yogaValues.setProperty(rt, key, value);
               }
             } catch (std::invalid_argument e) {
-              throw std::runtime_error("Failed to convert value to number");
+              throw std::runtime_error(
+                  "[Reanimated] Failed to convert value to number");
```
Comment (tomekzaw) [inline android/src/main/cpp/NativeProxy.cpp:511]:
```suggestion
                  "[Reanimated] Failed to convert value to number.");
```

File: docs/docs/guides/troubleshooting.mdx [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+**Problem:**
+This issue happens when Reanimated fails to initialize its native side from JavaScript.
+
+**Solution:**
+
+1. If you recently installed or upgraded Reanimated, make sure to rebuild your app.
+2. Check if your platform is supported by Reanimated. Currently we support:
+   - Android
+   - iOS
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:36]:
```suggestion
   - iOS
   - macOS
```

File: docs/docs/guides/troubleshooting.mdx [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+   - Web
+3. If you are running in a brownfield app, make sure to initialize the native library manually.
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:38]:
```suggestion
3. If you are using Reanimated in a brownfield app, make sure to initialize the native library manually.
```

File: docs/docs/guides/troubleshooting.mdx [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+### Failed to create a worklet
+
+**Problem:** This usually happens when Reanimated is not properly installed, e.g. forgetting to include the Reanimated Babel plugin in `babel.config.js`.
+
+**Solution:** See installation docs at https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation#babel-plugin for more information.
+
+### Native part of Reanimated doesn't seem to be initialized
+
+**Problem:**
+This issue happens when Reanimated fails to initialize its native side from JavaScript.
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:29]:
```suggestion
**Problem:** This issue happens when Reanimated fails to initialize its native side from JavaScript.
```

File: docs/docs/guides/troubleshooting.mdx [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+}
+```
+
+After that make sure to run your package manager again, either `yarn` or `npm install`.
+
+### Outdated version of `react-native` for New Architecture
+
+**Problem:** Reanimated supports New Architecture (Fabric) only on the latest minor release of React Native.
+
+**Solution:** Please upgrade to at React Native 0.72 or downgrade to an older version of Reanimated 3.
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:103]:
```suggestion
**Solution:** Please upgrade to at least React Native 0.72 or downgrade to an older version of Reanimated 3.
```

File: docs/docs/guides/troubleshooting.mdx [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+**Solution:**
+
+1. If you want to synchronously execute this method, mark it as a worklet using `worklet` directive:
+
+```diff
+function callee(){
++   'worklet';
+    console.log("hello");
+  }
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:132]:
```suggestion
 function callee() {
+  'worklet';
   console.log("hello");
 }
```

File: docs/docs/guides/troubleshooting.mdx [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+```
+
+2. If you want to execute this function on the JS thread, wrap it using `runOnJS`:
+
+```diff
+function caller(){
+  'worklet';
+-   callee();
++   runOnJS(callee)();
+}
```
Comment (tomekzaw) [inline docs/docs/guides/troubleshooting.mdx:142]:
```suggestion
 function caller() {
   'worklet';
-  callee();
+  runOnJS(callee)();
 }
```

File: scripts/reanimated_utils.rb [hunk 24]
```diff
@@ -27,7 +27,7 @@ def find_config()
   end
 
   if react_native_json == nil
-    raise '[RNReanimated] Unable to recognize your `react-native` version! Please set environmental variable with `react-native` locations: `export REACT_NATIVE_NODE_MODULES_DIR="<path to react-native>" && pod install'
+    raise '[Reanimated] Unable to recognize your `react-native` version! Please set environmental variable with `react-native` locations: `export REACT_NATIVE_NODE_MODULES_DIR="<path to react-native>" && pod install.'
```
Comment (tomekzaw) [inline scripts/reanimated_utils.rb:30]:
```suggestion
    raise '[Reanimated] Unable to recognize react-native version! Please set environmental variable with react-native location: `export REACT_NATIVE_NODE_MODULES_DIR="<path to react-native>" && pod install`.'
```

File: scripts/reanimated_utils.rb [hunk 25]
```diff
@@ -86,6 +86,6 @@ def assert_latest_react_native_with_new_architecture(config, reanimated_package_
   react_native_minor_version = config[:react_native_minor_version]
   fabric_enabled = ENV['RCT_NEW_ARCH_ENABLED'] == '1'
   if fabric_enabled && reanimated_major_version == 3 && react_native_minor_version < 72
-    raise "[react-native-reanimated] Reanimated " + reanimated_version + " supports the New Architecture only on the latest minor release of React Native. Please upgrade to React Native 0.72.0+ or downgrade to an older version of Reanimated v3"
+    raise "[Reanimated] Outdated version of `react-native` for New Architecture. Reanimated " + reanimated-version + " supports the New Architecture on React Native 0.72.0+."
```
Comment (tomekzaw) [inline scripts/reanimated_utils.rb:89]: Let's add a link to Troubleshooting section here.

File: scripts/reanimated_utils.rb [hunk 26]
```diff
@@ -68,15 +68,15 @@ def assert_no_multiple_instances(react_native_info)
...
+    raise "[Reanimated] Multiple versions of Reanimated were detected (iOS). See https://docs.swmansion.com/react-native-reanimated/docs/guides/Troubleshooting#multiple-version-of-reanimated-were-detected \n\nConflict between: \n" + parsed_location
   end
 end
 
 def assert_no_reanimated2_with_new_architecture(reanimated_package_json)
   reanimated_major_version = reanimated_package_json['version'].split('.')[0].to_i
   fabric_enabled = ENV['RCT_NEW_ARCH_ENABLED'] == '1'
   if fabric_enabled && reanimated_major_version == 2
-    raise "[react-native-reanimated] Reanimated 2.x does not support Fabric. Please upgrade to 3.x to use Reanimated with the New Architecture. For details, see https://blog.swmansion.com/announcing-reanimated-3-16167428c5f7"
+    raise "[Reanimated] Reanimated 2 does not support Fabric. Please upgrade to Reanimated 3 for the New Architecture support. For details, see https://blog.swmansion.com/announcing-reanimated-3-16167428c5f7."
```
Comment (tomekzaw) [inline scripts/reanimated_utils.rb:79]: Let's add a link to Troubleshooting section here.

File: src/reanimated2/interpolation.ts [hunk 27]
```diff
@@ -78,7 +78,7 @@ function validateType(type: ExtrapolationType): RequiredExtrapolationConfig {
   if (typeof type === 'string') {
     if (!isExtrapolate(type)) {
       throw new Error(
-        `Reanimated: not supported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
+        `[Reanimated] Unupported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```
Comment (tomekzaw) [inline src/reanimated2/interpolation.ts:81]:
```suggestion
        `[Reanimated] Unsupported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```

File: src/reanimated2/interpolation.ts [hunk 28]
```diff
@@ -93,7 +93,7 @@ function validateType(type: ExtrapolationType): RequiredExtrapolationConfig {
     (type.extrapolateRight && !isExtrapolate(type.extrapolateRight))
   ) {
     throw new Error(
-      `Reanimated: not supported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
+      `[Reanimated] Unupported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```
Comment (tomekzaw) [inline src/reanimated2/interpolation.ts:96]:
```suggestion
      `[Reanimated] Unsupported value for "interpolate" \nSupported values: ["extend", "clamp", "identity", Extrapolatation.CLAMP, Extrapolatation.EXTEND, Extrapolatation.IDENTITY]\n Valid example:
```

File: Common/cpp/Tools/SingleInstanceChecker.h [hunk 29]
```diff
@@ -28,7 +28,7 @@ class SingleInstanceChecker {
     if (!condition) {
 #ifdef ANDROID
       __android_log_print(
-          ANDROID_LOG_WARN, "Reanimated", "%s", message.c_str());
+          ANDROID_LOG_WARN, "[Reanimated] ", "%s", message.c_str());
```
[the remark this answers is not in the document]
Comment (tjzel) [reply Common/cpp/Tools/SingleInstanceChecker.h:31]: Sure, we can leave it as is.

File: android/build.gradle (hunk 15 above)
Comment (tjzel) [reply android/build.gradle:259]: Actually a few lines below this syntax is used, so I made the change for consistency there.

File: android/build.gradle (hunk 16 above)
Comment (tjzel) [reply android/build.gradle:445]: I think originally I wanted to give a link here but then I couldn't decide, definitely we can.

File: src/reanimated2/interpolation.ts (hunk 27 above)
Comment (tjzel) [reply src/reanimated2/interpolation.ts:81]: :grief:

**Review by tomekzaw — APPROVED** [review]

**Review by tomekzaw — CHANGES_REQUESTED** [review]
CI is red 😢

**Review by tomekzaw — APPROVED** [review]
CI is green 😄
